File: 206-ReverseLinkedList/206-ReverseLinkedList.java
// Last updated: 9/14/2025, 9:15:43 AM
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        //approach using recursion
        if(head == null || head.next == null) return head;

        ListNode newHead = reverseList(head.next);
        ListNode front = head.next;
        front.next = head;
        head.next = null;
        return newHead; 
    }
}

File: 206-ReverseLinkedList/206-ReverseLinkedList.c
// Last updated: 9/14/2025, 9:15:47 AM
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverseList(struct ListNode* head){
       struct ListNode *prev=NULL,*curr=head,*nxt;
       while(curr!=NULL){
           nxt=curr->next;
           curr->next=prev;
           prev=curr;
           curr=nxt;
       }
       return prev;
}
// 1 -> 2 -> 3 -> 4 -> NULL
// NULL <- 1 <- 2 <- 3 <- 4